File: backend/src/main/java/br/com/calciolari/datahub/imports/infrastructure/interpdv/qrp/InterPdvQrpParser.java
package br.com.calciolari.datahub.imports.infrastructure.interpdv.qrp;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import br.com.calciolari.datahub.imports.domain.parser.ImportParser;
import br.com.calciolari.datahub.imports.domain.parser.IssueSeverity;
import br.com.calciolari.datahub.imports.domain.parser.IssueStage;
import br.com.calciolari.datahub.imports.domain.parser.ParseIssue;
import br.com.calciolari.datahub.imports.domain.parser.ParsedImport;
import br.com.calciolari.datahub.imports.domain.parser.ParsedImportStats;
import br.com.calciolari.datahub.imports.domain.parser.ParsedImportTotals;
import br.com.calciolari.datahub.imports.domain.parser.ParserInput;
import br.com.calciolari.datahub.imports.domain.parser.SourceLocator;

/**
 * InterPDV QuickReport (.QRP) adapter. Binary layout ported from
 * {@code docs/poc/index.html} — not invented.
 */
public final class InterPdvQrpParser implements ImportParser {

	public static final String PARSER_NAME = InterPdvReportLayoutMapper.PARSER_NAME;
	public static final String PARSER_VERSION = InterPdvReportLayoutMapper.PARSER_VERSION;

	private final QrpContainerReader containerReader;
	private final EmfTextRecordExtractor textExtractor;
	private final InterPdvReportLayoutMapper layoutMapper;
	private final InterPdvParsedImportValidator validator;
	private final long maxBytes;

	public InterPdvQrpParser() {
		this(new QrpContainerReader(), new EmfTextRecordExtractor(), new InterPdvReportLayoutMapper(),
				new InterPdvParsedImportValidator(), 32L * 1024 * 1024);
	}

	public InterPdvQrpParser(
			QrpContainerReader containerReader,
			EmfTextRecordExtractor textExtractor,
			InterPdvReportLayoutMapper layoutMapper,
			InterPdvParsedImportValidator validator,
			long maxBytes) {
		this.containerReader = Objects.requireNonNull(containerReader);
		this.textExtractor = Objects.requireNonNull(textExtractor);
		this.layoutMapper = Objects.requireNonNull(layoutMapper);
		this.validator = Objects.requireNonNull(validator);
		if (maxBytes < 1) {
			throw new IllegalArgumentException("maxBytes must be >= 1");
		}
		this.maxBytes = maxBytes;
	}

	@Override
	public boolean supports(ParserInput input) {
		Objects.requireNonNull(input, "input");
		// Content sniffing happens in parse(); filename is never authoritative.
		// Provisional: accept when caller already classified as QRP or length looks plausible.
		if (input.detectedType() != null && input.detectedType().equalsIgnoreCase("QRP")) {
			return true;
		}
		String name = input.originalFilename();
		return name != null && name.toLowerCase().endsWith(".qrp");
	}

	@Override
	public ParsedImport parse(ParserInput input) {
		Objects.requireNonNull(input, "input");
		byte[] bytes = readLimited(input.content(), input.contentLength());
		List<EmfPage> pages = containerReader.findEmfPages(bytes);
		if (pages.isEmpty()) {
			return fatal("NO_EMF_PAGES", IssueStage.CONTAINER, "Nenhuma página EMF reconhecida neste QRP.");
		}

		List<EmfTextRun> texts = new ArrayList<>();
		for (int i = 0; i < pages.size(); i++) {
			texts.addAll(textExtractor.extract(bytes, pages.get(i), i));
		}
		if (texts.isEmpty()) {
			return fatal("NO_EMF_TEXT", IssueStage.EMF, "QRP reconhecido sem registros de texto EMF.");
		}

		ParsedImport mapped = layoutMapper.map(texts, pages.size());
		List<ParseIssue> merged = new ArrayList<>(mapped.issues());
		merged.addAll(validator.validate(mapped));
		return new ParsedImport(
				mapped.source(),
				mapped.parserName(),
				mapped.parserVersion(),
				mapped.externalProductId(),
				mapped.productName(),
				mapped.movements(),
				mapped.totals(),
				mapped.stats(),
				merged);
	}

	private byte[] readLimited(InputStream content, long contentLength) {
		try {
			if (contentLength > maxBytes) {
				throw new IllegalArgumentException("contentLength exceeds maxBytes (" + maxBytes + ")");
			}
			byte[] bytes = content.readAllBytes();
			if (bytes.length > maxBytes) {
				throw new IllegalArgumentException("payload exceeds maxBytes (" + maxBytes + ")");
			}
			if (contentLength > 0 && contentLength != bytes.length) {
				// Prefer actual bytes; length mismatch is non-fatal for parse itself.
			}
			return bytes;
		}
		catch (IOException ex) {
			throw new UncheckedIOException("failed to read parser input", ex);
		}
	}

	private static ParsedImport fatal(String code, IssueStage stage, String message) {
		return new ParsedImport(
				InterPdvReportLayoutMapper.SOURCE,
				PARSER_NAME,
				PARSER_VERSION,
				null,
				null,
				List.of(),
				ParsedImportTotals.empty(),
				ParsedImportStats.empty(),
				List.of(new ParseIssue(code, IssueSeverity.FATAL, stage, SourceLocator.empty(), message)));
	}
}

File: backend/src/Calciolari.DataHub/Imports/Infrastructure/InterPdv/Qrp/InterPdvQrpParser.cs
using Calciolari.DataHub.Imports.Domain.Parser;

namespace Calciolari.DataHub.Imports.Infrastructure.InterPdv.Qrp;

/// <summary>
/// InterPDV QuickReport (.QRP) adapter. Binary layout ported from docs/poc/index.html — not invented.
/// </summary>
public sealed class InterPdvQrpParser : IImportParser
{
    public const string ParserName = InterPdvReportLayoutMapper.ParserName;
    public const string ParserVersion = InterPdvReportLayoutMapper.ParserVersion;

    private readonly QrpContainerReader _containerReader;
    private readonly EmfTextRecordExtractor _textExtractor;
    private readonly InterPdvReportLayoutMapper _layoutMapper;
    private readonly InterPdvParsedImportValidator _validator;
    private readonly long _maxBytes;

    public InterPdvQrpParser()
        : this(new QrpContainerReader(), new EmfTextRecordExtractor(), new InterPdvReportLayoutMapper(),
            new InterPdvParsedImportValidator(), 32L * 1024 * 1024)
    {
    }

    public InterPdvQrpParser(
        QrpContainerReader containerReader,
        EmfTextRecordExtractor textExtractor,
        InterPdvReportLayoutMapper layoutMapper,
        InterPdvParsedImportValidator validator,
        long maxBytes)
    {
        _containerReader = containerReader ?? throw new ArgumentNullException(nameof(containerReader));
        _textExtractor = textExtractor ?? throw new ArgumentNullException(nameof(textExtractor));
        _layoutMapper = layoutMapper ?? throw new ArgumentNullException(nameof(layoutMapper));
        _validator = validator ?? throw new ArgumentNullException(nameof(validator));
        if (maxBytes < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(maxBytes), "maxBytes must be >= 1");
        }

        _maxBytes = maxBytes;
    }

    public bool Supports(ParserInput input)
    {
        ArgumentNullException.ThrowIfNull(input);
        if (input.DetectedType is not null &&
            string.Equals(input.DetectedType, "QRP", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        var name = input.OriginalFilename;
        return name is not null && name.EndsWith(".qrp", StringComparison.OrdinalIgnoreCase);
    }

    public ParsedImport Parse(ParserInput input)
    {
        ArgumentNullException.ThrowIfNull(input);
        var bytes = ReadLimited(input.Content, input.ContentLength);
        var pages = _containerReader.FindEmfPages(bytes);
        if (pages.Count == 0)
        {
            return Fatal("NO_EMF_PAGES", IssueStage.Container, "Nenhuma página EMF reconhecida neste QRP.");
        }

        var texts = new List<EmfTextRun>();
        for (var i = 0; i < pages.Count; i++)
        {
            texts.AddRange(_textExtractor.Extract(bytes, pages[i], i));
        }

        if (texts.Count == 0)
        {
            return Fatal("NO_EMF_TEXT", IssueStage.Emf, "QRP reconhecido sem registros de texto EMF.");
        }

        var mapped = _layoutMapper.Map(texts, pages.Count);
        var merged = mapped.Issues.ToList();
        merged.AddRange(_validator.Validate(mapped));
        return new ParsedImport(
            mapped.Source,
            mapped.ParserName,
            mapped.ParserVersion,
            mapped.ExternalProductId,
            mapped.ProductName,
            mapped.Movements,
            mapped.Totals,
            mapped.Stats,
            merged);
    }

    private byte[] ReadLimited(Stream content, long contentLength)
    {
        try
        {
            if (contentLength > _maxBytes)
            {
                throw new ArgumentException("contentLength exceeds maxBytes (" + _maxBytes + ")");
            }

            using var buffer = new MemoryStream();
            content.CopyTo(buffer);
            var bytes = buffer.ToArray();
            if (bytes.Length > _maxBytes)
            {
                throw new ArgumentException("payload exceeds maxBytes (" + _maxBytes + ")");
            }

            return bytes;
        }
        catch (IOException ex)
        {
            throw new IOException("failed to read parser input", ex);
        }
    }

    private static ParsedImport Fatal(string code, IssueStage stage, string message) =>
        new(
            InterPdvReportLayoutMapper.Source,
            ParserName,
            ParserVersion,
            null,
            null,
            Array.Empty<ParsedMovement>(),
            ParsedImportTotals.Empty,
            ParsedImportStats.Empty,
            [new ParseIssue(code, IssueSeverity.Fatal, stage, SourceLocator.Empty, message)]);
}
